From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Thu, 24 Jul 2025 14:07:47 +0300
Subject: [PATCH] Optimize level ticking


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 734cbfc6dce55da25af27aa75dae8fa3c3cf043a..45e8d83590c2c77a2e0d154e20841d68e1339817 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1037,9 +1037,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     // Paper start - optimise random ticking
     private final ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom simpleRandom = new ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom(net.minecraft.world.level.levelgen.RandomSupport.generateUniqueSeed());
 
+    // DivineMC start - Optimize level ticking
     private void optimiseRandomTick(final LevelChunk chunk, final int tickSpeed) {
         final LevelChunkSection[] sections = chunk.getSections();
-        final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection((ServerLevel)(Object)this);
+        final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection(this); // DivineMC - Optimize level ticking
         final ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom simpleRandom = this.simpleRandom;
         final boolean doubleTickFluids = !ca.spottedleaf.moonrise.common.PlatformHooks.get().configFixMC224294();
 
@@ -1048,42 +1049,38 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         final int offsetZ = cpos.z() << 4;
 
         for (int sectionIndex = 0, sectionsLen = sections.length; sectionIndex < sectionsLen; sectionIndex++) {
-            final int offsetY = (sectionIndex + minSection) << 4;
             final LevelChunkSection section = sections[sectionIndex];
+            if (!section.isRandomlyTickingBlocks()) continue;
+            final int offsetY = (sectionIndex + minSection) << 4;
             final net.minecraft.world.level.chunk.PalettedContainer<net.minecraft.world.level.block.state.BlockState> states = section.states;
-            if (!section.isRandomlyTickingBlocks()) {
-                continue;
-            }
 
-            final ca.spottedleaf.moonrise.common.list.ShortList tickList = ((ca.spottedleaf.moonrise.patches.block_counting.BlockCountingChunkSection)section).moonrise$getTickingBlockList();
+            final ca.spottedleaf.moonrise.common.list.ShortList tickList = section.moonrise$getTickingBlockList();
 
             for (int i = 0; i < tickSpeed; ++i) {
-                final int tickingBlocks = tickList.size();
                 final int index = simpleRandom.nextInt() & ((16 * 16 * 16) - 1);
 
-                if (index >= tickingBlocks) {
+                if (index >= tickList.size()) {
                     // most of the time we fall here
                     continue;
                 }
 
-                final int location = (int)tickList.getRaw(index) & 0xFFFF;
+                final int location = tickList.getRaw(index);
                 final BlockState state = states.get(location);
 
                 // do not use a mutable pos, as some random tick implementations store the input without calling immutable()!
-                final BlockPos pos = new BlockPos((location & 15) | offsetX, ((location >>> (4 + 4)) & 15) | offsetY, ((location >>> 4) & 15) | offsetZ);
+                final BlockPos pos = new BlockPos((location & 15) | offsetX, (location >>> (4 + 4)) | offsetY, ((location >>> 4) & 15) | offsetZ);
 
-                state.randomTick((ServerLevel)(Object)this, pos, simpleRandom);
+                state.randomTick(this, pos, simpleRandom);
                 if (doubleTickFluids) {
                     final FluidState fluidState = state.getFluidState();
                     if (fluidState.isRandomlyTicking()) {
-                        fluidState.randomTick((ServerLevel)(Object)this, pos, simpleRandom);
+                        fluidState.randomTick(this, pos, simpleRandom);
                     }
                 }
             }
         }
-
-        return;
     }
+    // DivineMC end - Optimize level ticking
     // Paper end - optimise random ticking
 
     public void tickChunk(final LevelChunk chunk, final int tickSpeed) {
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 24a880c588c31d3f2e24e1fbf366789095e62fb8..cf70bce6dd6870faf4f49fa3a23323ebb6f39591 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -86,7 +86,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
             return "<null>";
         }
     };
-    private final Map<BlockPos, LevelChunk.RebindableTickingBlockEntityWrapper> tickersInLevel = Maps.newHashMap();
+    private final Map<BlockPos, LevelChunk.RebindableTickingBlockEntityWrapper> tickersInLevel = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(); // DivineMC - Optimize level ticking
     public boolean loaded;
     public final ServerLevel level; // CraftBukkit - type
     private @Nullable Supplier<FullChunkStatus> fullStatus;
diff --git a/net/minecraft/world/ticks/LevelChunkTicks.java b/net/minecraft/world/ticks/LevelChunkTicks.java
index c3825e91504f5df6bfd8e41de1f5dc4ae45d81e2..6f14a4e359b74b8b0e7177eee2edd38d3c3bc71a 100644
--- a/net/minecraft/world/ticks/LevelChunkTicks.java
+++ b/net/minecraft/world/ticks/LevelChunkTicks.java
@@ -14,9 +14,9 @@ import net.minecraft.core.BlockPos;
 import org.jspecify.annotations.Nullable;
 
 public class LevelChunkTicks<T> implements TickContainerAccess<T>, SerializableTickContainer<T>, ca.spottedleaf.moonrise.patches.chunk_system.ticks.ChunkSystemLevelChunkTicks { // Paper - rewrite chunk system
-    private final Queue<ScheduledTick<T>> tickQueue = new PriorityQueue<>(ScheduledTick.DRAIN_ORDER);
+    private final Queue<ScheduledTick<T>> tickQueue = new java.util.concurrent.PriorityBlockingQueue<>(11, ScheduledTick.DRAIN_ORDER); // DivineMC - Optimize level ticking
     private @Nullable List<SavedTick<T>> pendingTicks;
-    private final Set<ScheduledTick<?>> ticksPerPosition = new ObjectOpenCustomHashSet<>(ScheduledTick.UNIQUE_TICK_HASH);
+    private final Set<ScheduledTick<?>> ticksPerPosition = it.unimi.dsi.fastutil.objects.ObjectSets.synchronize(new ObjectOpenCustomHashSet<>(ScheduledTick.UNIQUE_TICK_HASH)); // DivineMC - Optimize level ticking
     private @Nullable BiConsumer<LevelChunkTicks<T>, ScheduledTick<T>> onTickAdded;
 
     // Paper start - rewrite chunk system
@@ -63,10 +63,18 @@ public class LevelChunkTicks<T> implements TickContainerAccess<T>, SerializableT
     }
 
     public @Nullable ScheduledTick<T> poll() {
-        ScheduledTick<T> result = this.tickQueue.poll();
-        if (result != null) {
-            this.ticksPerPosition.remove(result); this.dirty = true; // Paper - rewrite chunk system
+        // DivineMC start - Optimize collections
+        ScheduledTick<T> result = null;
+        try {
+            result = this.tickQueue.poll();
+            if (result != null) {
+                this.ticksPerPosition.remove(result); this.dirty = true; // Paper - rewrite chunk system
+            }
+        } catch (Exception e) {
+            net.minecraft.server.MinecraftServer.LOGGER.error("Encountered caught exception when polling chunk ticks, blocking and returning null.", e);
+            return null;
         }
+        // DivineMC end - Optimize collections
 
         return result;
     }
@@ -79,6 +87,7 @@ public class LevelChunkTicks<T> implements TickContainerAccess<T>, SerializableT
     }
 
     private void scheduleUnchecked(final ScheduledTick<T> tick) {
+        if (tick == null) return; // DivineMC - Optimize level ticking
         this.tickQueue.add(tick);
         if (this.onTickAdded != null) {
             this.onTickAdded.accept(this, tick);
@@ -120,6 +129,7 @@ public class LevelChunkTicks<T> implements TickContainerAccess<T>, SerializableT
         }
 
         for (ScheduledTick<T> tick : this.tickQueue) {
+            if (tick == null) continue; // DivineMC - Optimize level ticking
             ticks.add(tick.toSavedTick(currentTick));
         }
 
diff --git a/net/minecraft/world/ticks/LevelTicks.java b/net/minecraft/world/ticks/LevelTicks.java
index 7bf9e93692debcef511bfa316aeef32ae8a7d3b7..00707ad2b960bd3dd5dcc668e28ae753835722e0 100644
--- a/net/minecraft/world/ticks/LevelTicks.java
+++ b/net/minecraft/world/ticks/LevelTicks.java
@@ -31,10 +31,12 @@ import net.minecraft.world.level.levelgen.structure.BoundingBox;
 public class LevelTicks<T> implements LevelTickAccess<T> {
     private static final Comparator<LevelChunkTicks<?>> CONTAINER_DRAIN_ORDER = (o1, o2) -> ScheduledTick.INTRA_TICK_DRAIN_ORDER.compare(o1.peek(), o2.peek());
     private final LongPredicate tickCheck;
-    private final Long2ObjectMap<LevelChunkTicks<T>> allContainers = new Long2ObjectOpenHashMap<>();
-    private final Long2LongMap nextTickForContainer = Util.make(new Long2LongOpenHashMap(), m -> m.defaultReturnValue(Long.MAX_VALUE));
-    private final Queue<LevelChunkTicks<T>> containersToTick = new PriorityQueue<>(CONTAINER_DRAIN_ORDER);
-    private final Queue<ScheduledTick<T>> toRunThisTick = new ArrayDeque<>();
+    // DivineMC start - Optimize collections
+    private final Long2ObjectMap<LevelChunkTicks<T>> allContainers = it.unimi.dsi.fastutil.longs.Long2ObjectMaps.synchronize(new Long2ObjectOpenHashMap<>());
+    private final java.util.Map<Long, Long> nextTickForContainer = new java.util.concurrent.ConcurrentHashMap<>();
+    private final Queue<LevelChunkTicks<T>> containersToTick = new java.util.concurrent.PriorityBlockingQueue<>(11, CONTAINER_DRAIN_ORDER);
+    private final Queue<ScheduledTick<T>> toRunThisTick = new java.util.concurrent.ConcurrentLinkedQueue<>();
+    // DivineMC end - Optimize collections
     private final List<ScheduledTick<T>> alreadyRunThisTick = new ArrayList<>();
     private final Set<ScheduledTick<?>> toRunThisTickSet = new ObjectOpenCustomHashSet<>(ScheduledTick.UNIQUE_TICK_HASH);
     private final BiConsumer<LevelChunkTicks<T>, ScheduledTick<T>> chunkScheduleUpdater = (container, newTick) -> {
@@ -42,6 +44,7 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
             this.updateContainerScheduling(newTick);
         }
     };
+    private final java.util.concurrent.atomic.AtomicInteger toRunThisTickCount = new java.util.concurrent.atomic.AtomicInteger(0); // DivineMC - Optimize level ticking
 
     public LevelTicks(final LongPredicate tickCheck) {
         this.tickCheck = tickCheck;
@@ -98,12 +101,14 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
     }
 
     private void sortContainersToTick(final long currentTick) {
-        ObjectIterator<Entry> it = Long2LongMaps.fastIterator(this.nextTickForContainer);
+        java.util.Iterator<java.util.Map.Entry<Long, Long>> it = this.nextTickForContainer.entrySet().iterator(); // DivineMC - Optimize level ticking
 
         while (it.hasNext()) {
-            Entry entry = it.next();
-            long chunkPos = entry.getLongKey();
-            long nextTick = entry.getLongValue();
+            // DivineMC start - Optimize collections
+            java.util.Map.Entry<Long, Long> entry = it.next();
+            long chunkPos = entry.getKey();
+            long nextTick = entry.getValue();
+            // DivineMC end - Optimize collections
             if (nextTick <= currentTick) {
                 LevelChunkTicks<T> candidateContainer = this.allContainers.get(chunkPos);
                 if (candidateContainer == null) {
@@ -172,16 +177,19 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
     }
 
     private void scheduleForThisTick(final ScheduledTick<T> tick) {
+        if (tick == null) return; // DivineMC - Optimize level ticking
         this.toRunThisTick.add(tick);
+        this.toRunThisTickCount.incrementAndGet(); // DivineMC - Optimize level ticking
     }
 
     private boolean canScheduleMoreTicks(final int maxTicksToProcess) {
-        return this.toRunThisTick.size() < maxTicksToProcess;
+        return this.toRunThisTickCount.get() < maxTicksToProcess; // DivineMC - Optimize level ticking
     }
 
     private void runCollectedTicks(final BiConsumer<BlockPos, T> output) {
         while (!this.toRunThisTick.isEmpty()) {
             ScheduledTick<T> entry = this.toRunThisTick.poll();
+            this.toRunThisTickCount.decrementAndGet(); // DivineMC - Optimize level ticking
             if (!this.toRunThisTickSet.isEmpty()) {
                 this.toRunThisTickSet.remove(entry);
             }
@@ -192,7 +200,7 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
     }
 
     private void cleanupAfterTick() {
-        this.toRunThisTick.clear();
+        this.toRunThisTickCount.set(0); // DivineMC - Optimize level ticking
         this.containersToTick.clear();
         this.alreadyRunThisTick.clear();
         this.toRunThisTickSet.clear();
