From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dan28000 <84990628+dan28000@users.noreply.github.com>
Date: Sun, 31 May 2026 16:05:34 +0200
Subject: [PATCH] t


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 7ed81ec2ed25ec8feb211d3d14bf29d4e6ff5a1b..b7c83bcd565aae95d02cfeccdd2d1d5c133c3d7b 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -459,7 +459,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private boolean ramBar = false; // Purpur - Implement rambar commands
     public boolean hasTickedAtLeastOnceInNewWorld = false; // DivineMC - Parallel world ticking
     public final org.bxteam.divinemc.async.rct.RollingLongBuffer avgTickTimeNanos = new org.bxteam.divinemc.async.rct.RollingLongBuffer(100); // DivineMC - Region tick time tracking
-    public int lastRegionChunkSize, lastRegionEntityAmount = 0;
+    public int lastRegionChunkSize, lastRegionEntityAmount, regionHash = 0;
 
     // Paper start - rewrite chunk system
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
